README.md
# MCE Table Buttons for WordPress

> Adds table editing controls to the visual content editor (TinyMCE).

[![Support Level](https://img.shields.io/badge/support-stable-blue.svg)](#support-level) [![Linting](https://github.com/10up/mce-table-buttons/actions/workflows/lint.yml/badge.svg)](https://github.com/10up/mce-table-buttons/actions/workflows/lint.yml) [![Release Version](https://img.shields.io/github/release/10up/mce-table-buttons.svg)](https://github.com/10up/mce-table-buttons/releases/latest) ![WordPress tested up to version](https://img.shields.io/wordpress/plugin/tested/mce-table-buttons?label=WordPress) [![GPLv2 License](https://img.shields.io/github/license/10up/mce-table-buttons.svg)](https://github.com/10up/mce-table-buttons/blob/develop/LICENSE.md)

## Overview

A light weight plug-in that adds the table editing controls from the full version of TinyMCE, optimized for WordPress. Note that this may not work in conjunction with other plug-ins that significantly alter or replace the visual editor's default behavior.

Note that the table controls are contained in the “kitchen sink” toolbar, toggled with the last button on the first row of controls.

![Screenshot of added table buttons](.wordpress-org/screenshot-1.png)

## Technical information

* Requires at least WordPress 4.0
* Includes TinyMCE table plugin versions 4.7.11 and 4.1.?

## Installation

1. Install easily with the WordPress plugin control panel or manually download the plugin and upload the folder
`mce-table-buttons` to the `/wp-content/plugins/` directory
1. Activate the plugin through the 'Plugins' menu in WordPress
1. Show the toolbar in the editor by opening the "kitchen sink" (the last button in the first row)

## Support Level

**Stable:** 10up is not planning to develop any new features for this, but will still respond to bug reports and security concerns.  We welcome PRs, but any that include new features should be small and easy to integrate and should not include breaking changes. We otherwise intend to keep this tested up to the most recent version of WordPress.

## Changelog

A complete listing of all notable changes to MCE Table Buttons are documented in [CHANGELOG.md](https://github.com/10up/mce-table-buttons/blob/develop/CHANGELOG.md).

## Contributing

Please read [CODE_OF_CONDUCT.md](https://github.com/10up/mce-table-buttons/blob/develop/CODE_OF_CONDUCT.md) for details on our code of conduct, [CONTRIBUTING.md](https://github.com/10up/mce-table-buttons/blob/develop/CONTRIBUTING.md) for details on the process for submitting pull requests to us, and [CREDITS.md](https://github.com/10up/mce-table-buttons/blob/develop/CREDITS.md) for a listing of maintainers of, contributors to, and libraries used by MCE Table Buttons.

## Like what you see?

<p align="center">
<a href="http://10up.com/contact/"><img src="https://10up.com/uploads/2016/10/10up-Github-Banner.png" width="850"></a>
</p>


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file, per [the Keep a Changelog standard](http://keepachangelog.com/).

## [Unreleased] - TBD

## [3.3.1] - 2022-01-26
**Note that this release replaces references of `master` with `trunk` so please plan to update any references you have from `master` to `trunk` to ensure you continue to receive the latest updates on this plugin.**

### Changed
- Bump WordPress version “tested up to” 5.9 (props [David Chabbi](https://profiles.wordpress.org/davidchabbi/), [@jeffpaul](https://github.com/jeffpaul), [@ciprianimike](https://github.com/ciprianimike), [@sudip-10up](https://github.com/sudip-10up) via [#12](https://github.com/10up/mce-table-buttons/pull/12), [#18](https://github.com/10up/mce-table-buttons/pull/18), [#23](https://github.com/10up/mce-table-buttons/pull/23)).
- Updated documentation including new GitHub Actions (props [@jeffpaul](https://github.com/jeffpaul), [@dinhtungdu](https://github.com/dinhtungdu) via [#5](https://github.com/10up/mce-table-buttons/pull/5), [#13](https://github.com/10up/mce-table-buttons/pull/13), [#15](https://github.com/10up/mce-table-buttons/pull/15)).

### Fixed
- Issues reported by PHPCS WPCS v2.3.0 (props [@ciprianimike](https://github.com/ciprianimike), [@dinhtungdu](https://github.com/dinhtungdu) via [#17](https://github.com/10up/mce-table-buttons/pull/17)).

## [3.3] - 2018-06-15
### Changed
- Significantly update TinyMCE plugin from 4.1.x to 4.7.11
- Drop support for WordPress 3.x, which means dropping TinyMCE 3.x and 4.0.x

## [3.2] - 2014-08-24
### Changed
- WordPress 4.0 support, including a much newer TinyMCE plugin, with many new features, including background color and horizontal alignment
- Dramatically improved support for paragraphed content inside of a cell (paragraph breaks no longer disappear on save)

## [3.1] - 2014-05-31
### Changed
- Updated core TinyMCE table plugin from 4.0.20 to 4.0.21 in sync with WordPress - mostly bug and accessibility fixes
- Refactored for compatibility with plugins like Advanced Custom Fields that do not use the_editor hook

## [3.0] - 2014-04-01
### Changed
- Support for WordPress 3.9 and newer, which includes a major visual editor upgrade (TinyMCE 4)

## [2.0] - 2012-11-18
### Added
- New button icons that better conform with WordPress's editor design
- Retina (HiDPI) ready button icons!

### Changed
- Upgraded to latest version of TinyMCE tables plugin (fixes a lot of edge case bugs)
- Rewrote code for hiding / display toolbar with kitchen sink (now a TinyMCE plug-in instead of a workaround) - the table buttons no longer briefly appear before page loading is finished

## [1.5] - 2011-12-23
### Changed
- Table toolbar is hidden or displayed along with the kitchen sink (yay!)
- Minor clean up to code base and files; optimized for WordPress 3.3

## [1.0.4] - 2011-03-27
### Changed
- Updated TinyMCE table plug-in to corresponding TinyMCE update in WordPress 3.1 (still supports <3.1 too)

## [1.0.3.1] - 2011-02-22
### Changed
- Updated support / developer information

## [1.0.3] - 2011-01-17
### Changed
- Code clean up, 3.1 testing

## [1.0.2] - 2010-11-16
### Changed
- Minor code clean up

### Fixed
- Editor save warning appearing for unchanged content when the table is at the bottom of the editor

## [1.0.1] - 2009-12-08
### Fixed
- Issue with WebKit browsers (Safari and Chrome) - TinyMCE bug

## [1.0] - 2009-11-06
- Initial plugin release :tada:

[Unreleased]: https://github.com/10up/mce-table-buttons/compare/trunk...develop
[3.3.1]: https://github.com/10up/mce-table-buttons/compare/3.3...3.3.1
[3.3]: https://github.com/10up/mce-table-buttons/commit/7b1f57e
[3.2]: https://plugins.trac.wordpress.org/changeset/971857/
[3.1]: https://plugins.trac.wordpress.org/changeset/924344/
[3.0]: https://plugins.trac.wordpress.org/changeset/924344/
[2.0]: https://plugins.trac.wordpress.org/changeset/626531/
[1.5]: https://plugins.trac.wordpress.org/changeset/479678/
[1.0.4]: https://plugins.trac.wordpress.org/changeset/365544/
[1.0.3.1]: https://plugins.trac.wordpress.org/changeset/349281/
[1.0.3]: https://plugins.trac.wordpress.org/changeset/333541/
[1.0.2]: https://plugins.trac.wordpress.org/changeset/312235/
[1.0.1]: https://plugins.trac.wordpress.org/changeset/180734/
[1.0]: https://plugins.trac.wordpress.org/changeset/170960/


mce_table_buttons.php
<?php
/**
 * Plugin Name: MCE Table Buttons
 * Plugin URI: http://10up.com/plugins-modules/wordpress-mce-table-buttons/
 * Description: Add <strong>controls for table editing</strong> to the visual content editor with this <strong>light weight</strong> plug-in.
 * Version: 3.3
 * Author: Jake Goldman, 10up, Oomph
 * Author URI: http://10up.com
 * License: GPLv2 or later
 */

class MCE_Table_Buttons {

	/**
	 * Handles initializing this class and returning the singleton instance after it's been cached.
	 *
	 * @return null|MCE_Table_Buttons
	 */
	public static function get_instance() {
		// Store the instance locally to avoid private static replication
		static $instance = null;

		if ( null === $instance ) {
			$instance = new self();
			self::_setup_plugin();
		}

		return $instance;
	}

	/**
	 * An empty constructor
	 */
	public function __construct() { /* Purposely do nothing here */ }

	/**
	 * Handles registering hooks that initialize this plugin.
	 */
	public static function _setup_plugin() {
		add_filter( 'mce_external_plugins', array( __CLASS__, 'mce_external_plugins' ) );
		add_filter( 'mce_buttons_2', array( __CLASS__, 'mce_buttons_2' ) );
		add_filter( 'content_save_pre', array( __CLASS__, 'content_save_pre' ), 20 );
		add_filter( 'tiny_mce_before_init', array( __CLASS__, 'tiny_mce_before_init' ), 10, 2 );
	}

	/**
	 * Initialize TinyMCE table plugin and custom TinyMCE plugin
	 *
	 * @param array $plugin_array Array of TinyMCE plugins
	 * @return array Array of TinyMCE plugins
	 */
	public static function mce_external_plugins( $plugin_array ) {
		global $tinymce_version;
		$variant = ( defined('SCRIPT_DEBUG') && SCRIPT_DEBUG ) ? '' : '.min';

		if ( version_compare( $tinymce_version, '4700', '<' ) ) {

			$plugin_array['table'] = plugin_dir_url( __FILE__ ) . 'tinymce41-table/plugin' . $variant . '.js';

		} else {

			$plugin_array['table'] = plugin_dir_url( __FILE__ ) . 'tinymce47-table/plugin' . $variant . '.js';

		}

		return $plugin_array;
	}

	/**
	 * Add TinyMCE table control buttons
	 *
	 * @param array $buttons Buttons for the second row
	 * @return array Buttons for the second row
	 */
	public static function mce_buttons_2( $buttons ) {
		// in case someone is manipulating other buttons, drop table controls at the end of the row
		if ( ! $pos = array_search( 'undo', $buttons ) ) {
			array_push( $buttons, 'table' );
			return $buttons;
		}

		$buttons = array_merge( array_slice( $buttons, 0, $pos ), array( 'table' ), array_slice( $buttons, $pos ) );

		return $buttons;
	}

	/**
	 * Fixes weirdness resulting from wpautop and formatting clean up not built for tables
	 *
	 * @param string $content Editor content before WordPress massaging
	 * @return string Editor content before WordPress massaging
	 */
	public static function content_save_pre( $content ) {
		if ( false !== strpos( $content, '<table' ) ) {
			// paragraphed content inside of a td requires first paragraph to have extra line breaks (or else autop breaks)
			$content  = preg_replace( "/<td([^>]*)>(.+\r?\n\r?\n)/m", "<td$1>\n\n$2", $content );

			// make sure there's space around the table
			if ( substr( $content, -8 ) == '</table>' ) {
				$content .= "\n<br />";
			}
		}
		
		return $content;
	}

	/**
	 * Remove the table toolbar introduced in TinyMCE 4.3.0.
	 *
	 * Its positioning does not work correctly inside WordPress and blocks editing.
	 *
	 * @param array  $mceInit   An array with TinyMCE config.
	 * @param string $editor_id Unique editor identifier, e.g. 'content'.
	 *
	 * @return array TinyMCE config array
	 */
	public static function tiny_mce_before_init ( $mceInit, $editor_id ) {
		$mceInit['table_toolbar'] = '';

		return $mceInit;
	}
}

MCE_Table_Buttons::get_instance();

class-mce-table-buttons.php
<?php
/**
 * MCE Table Buttons
 *
 * @package           MCETableButtons
 * 
 * @wordpress-plugin
 * Plugin Name:       MCE Table Buttons
 * Plugin URI:        http://10up.com/plugins-modules/wordpress-mce-table-buttons/
 * Description:       Add <strong>controls for table editing</strong> to the visual content editor with this <strong>light weight</strong> plug-in.
 * Version:           3.3.1
 * Requires at least: 4.0
 * Requires PHP:      
 * Author:            Jake Goldman, 10up, Oomph
 * Author URI:        https://10up.com
 * License:           GPL v2 or later
 * License URI:       https://www.gnu.org/licenses/gpl-2.0.html
 */

/**
 * MCE Table Buttons main class.
 */
class MCE_Table_Buttons {
	/**
	 * Handles initializing this class and returning the singleton instance after it's been cached.
	 *
	 * @return null|MCE_Table_Buttons
	 */
	public static function get_instance() {
		// Store the instance locally to avoid private static replication.
		static $instance = null;

		if ( null === $instance ) {
			$instance = new self();
			self::setup_plugin();
		}

		return $instance;
	}

	/**
	 * An empty constructor
	 */
	public function __construct() {
		/* Purposely do nothing here. */
	}

	/**
	 * Handles registering hooks that initialize this plugin.
	 */
	public static function setup_plugin() {
		add_filter( 'mce_external_plugins', array( __CLASS__, 'mce_external_plugins' ) );
		add_filter( 'mce_buttons_2', array( __CLASS__, 'mce_buttons_2' ) );
		add_filter( 'content_save_pre', array( __CLASS__, 'content_save_pre' ), 20 );
		add_filter( 'tiny_mce_before_init', array( __CLASS__, 'tiny_mce_before_init' ), 10, 2 );
	}

	/**
	 * Initialize TinyMCE table plugin and custom TinyMCE plugin
	 *
	 * @param array $plugin_array Array of TinyMCE plugins.
	 * @return array Array of TinyMCE plugins
	 */
	public static function mce_external_plugins( $plugin_array ) {
		global $tinymce_version;
		$variant = ( defined( 'SCRIPT_DEBUG' ) && SCRIPT_DEBUG ) ? '' : '.min';

		if ( version_compare( $tinymce_version, '4700', '<' ) ) {

			$plugin_array['table'] = plugin_dir_url( __FILE__ ) . 'tinymce41-table/plugin' . $variant . '.js';

		} else {

			$plugin_array['table'] = plugin_dir_url( __FILE__ ) . 'tinymce47-table/plugin' . $variant . '.js';

		}

		return $plugin_array;
	}

	/**
	 * Add TinyMCE table control buttons
	 *
	 * @param array $buttons Buttons for the second row.
	 * @return array Buttons for the second row
	 */
	public static function mce_buttons_2( $buttons ) {
		// in case someone is manipulating other buttons, drop table controls at the end of the row.
		$pos = array_search( 'undo', $buttons, true );
		if ( ! $pos ) {
			array_push( $buttons, 'table' );
			return $buttons;
		}

		$buttons = array_merge( array_slice( $buttons, 0, $pos ), array( 'table' ), array_slice( $buttons, $pos ) );

		return $buttons;
	}

	/**
	 * Fixes weirdness resulting from wpautop and formatting clean up not built for tables
	 *
	 * @param string $content Editor content before WordPress massaging.
	 * @return string Editor content before WordPress massaging
	 */
	public static function content_save_pre( $content ) {
		if ( false !== strpos( $content, '<table' ) ) {
			// paragraphed content inside of a td requires first paragraph to have extra line breaks (or else autop breaks).
			$content = preg_replace( "/<td([^>]*)>(.+\r?\n\r?\n)/m", "<td$1>\n\n$2", $content );

			// make sure there's space around the table.
			if ( '</table>' === substr( $content, -8 ) ) {
				$content .= "\n<br />";
			}
		}

		return $content;
	}

	/**
	 * Remove the table toolbar introduced in TinyMCE 4.3.0.
	 *
	 * Its positioning does not work correctly inside WordPress and blocks editing.
	 *
	 * @param array  $mce_init   An array with TinyMCE config.
	 * @param string $editor_id Unique editor identifier, e.g. 'content'.
	 *
	 * @return array TinyMCE config array
	 */
	public static function tiny_mce_before_init( $mce_init, $editor_id ) {
		$mce_init['table_toolbar'] = '';

		return $mce_init;
	}
}

MCE_Table_Buttons::get_instance();
